File: leetcode/Contest/Weekly/222/1712.cpp
// 1712. Ways to Split Array Into Three Subarrays
// https://leetcode.com/problems/ways-to-split-array-into-three-subarrays/

class Solution {
public:
    int waysToSplit(vector<int>& nums) {
        int n = nums.size();
        int M = 1e9+7;
        
        vector<int> prefix(n, 0);
        prefix[0] = nums[0];
        for (int i = 1; i < n; i++)
            prefix[i] = prefix[i-1] + nums[i];
        
        long long res = 0;
        for (int i = 0; i < n - 1; i++){
            int left = prefix[i];
            int remaining = prefix[n-1] - left;
            int max_mid = remaining / 2;
            
            int mid_start = lower_bound(prefix.begin()+i+1, prefix.end() - 1, left*2) - prefix.begin();
            int right_start = upper_bound(prefix.begin()+i+1, prefix.end() - 1, left + max_mid) - prefix.begin();
            
            res += max(0, right_start - mid_start);
            res %= M;
        }
        
        return res % M;
    }
};

// O(1) space
class Solution {
public:
    int waysToSplit(vector<int>& prefix) {
        int n = prefix.size();
        int M = 1e9+7;
        
        for (int i = 1; i < n; i++)
            prefix[i] += prefix[i-1];
        
        long long res = 0;
        for (int i = 0; i < n - 1; i++){
            int left = prefix[i];
            int remaining = prefix[n-1] - left;
            int max_mid = remaining / 2;
            
            int mid_start = lower_bound(prefix.begin()+i+1, prefix.end() - 1, left*2) - prefix.begin();
            int right_start = upper_bound(prefix.begin()+i+1, prefix.end() - 1, left + max_mid) - prefix.begin();
            
            res += max(0, right_start - mid_start);
            res %= M;
        }
        
        return res % M;
    }
};

File: Leetcode Contest/Weekly/222/1712.py
# 1712. Ways to Split Array Into Three Subarrays
# https://leetcode.com/problems/ways-to-split-array-into-three-subarrays/

class Solution:
    def waysToSplit(self, nums: List[int]):
        prefix = [0]
        for num in nums:
            prefix.append(prefix[-1] + num)
        
        M = 10 ** 9 + 7
        ans = 0
        for i in range(1, len(nums)): 
            j = bisect_left(prefix, 2*prefix[i])
            k = bisect_right(prefix, (prefix[i] + prefix[-1])//2)
            ans += max(0, min(len(nums), k) - max(i+1, j))

        return ans % M
    
    def helper(self, prefix, leftSum, i, searchLeft):
        n = len(prefix)
        left, right = i, n - 2
        res = -1
        
        while left <= right:
            mid = left + (right-left) // 2
            
            midSum = prefix[mid] - prefix[i - 1]
            rightSum = prefix[-1] - prefix[mid]
            
            if leftSum <= midSum <= rightSum:
                res = mid
                if searchLeft: right = mid - 1
                else: left = mid + 1
            elif leftSum > midSum:
                left = mid + 1
            else:
                right = mid - 1
        
        return res
        
    def waysToSplit(self, nums: List[int]):
        
        M = 10 ** 9 + 7
        n = len(nums)
        prefix = [nums[0]]
        for num in nums[1:]:
            prefix.append(prefix[-1] + num)
        
        res = 0
        for i in range(1, n-1):
            leftSum = prefix[i-1]
            
            mid = self.helper(prefix, leftSum, i, True)
            right = self.helper(prefix, leftSum, i, False)
            
            if mid == -1 or right == - 1: continue
            
            res += (right - mid + 1) 
            res %= M
        
        return res % M
    